### DISCUSSION TITLE
How to fix virtual IP addresses instead of requesting them in client mode

### DISCUSSION THREAD

**Discussion by TelDragon** [strongswan/strongswan#2304] [Q&A] [answered]
How to fix virtual IP addresses instead of requesting them in client mode

My server is configured with a virtual address pool
```
pools {
    rw_pool {
      addrs = 11.11.11.0/24
    }
  }
```

The client is currently using a plea method to obtain any available IP address

```
vips = 0.0.0.0
```
I want to fix this virtual IP address.

Namely, client1 allocation 11.11.11.1
```
vips = 11.11.11.1
```

Client 2 allocation 11.11.11.2
```
vips = 11.11.11.2
```

I have reviewed the instructions about<conn>. vips

[https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html](https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html)

```
Comma-separated list of virtual IPs to request in IKEv2 configuration payloads or IKEv1 ModeConfig. The wildcard addresses 0.0.0.0 and :: request an arbitrary address, specific addresses may be defined. The responder may return a different address, though, or none at all
```

Of course, I chose IKEV2

The main reason is that my business backend needs to do net address translation.
I need to set a fixed private network virtual IP address

**Comment by tobiasbrunner** [comment, marked as answer]
You probably want to assign virtual IPs from a database or via RADIUS based on the client's identities (no change required on the clients). Please read the [documentation about virtual IPs](https://docs.strongswan.org/docs/5.9/features/vip.html) for details.

**Reply by TelDragon** [reply]
good At present, I have preliminarily implemented radius authentication and assigned client virtual IP addresses

Strongswan opens the eap-radius plugin section
```
vim /etc/strongswan/strongswan.d/charon/eap-radius.conf

accounting = yes
servers {
       freeradius {
          address = 172.17.62.205
          secret = test123
          auth_port = 1812
          acct_port = 1813

       }
    }
```

swanctl-conf section
```
        pools = radius
        remote {
#            auth = eap-mschapv2
            auth = eap-radius
            eap_id = %any
        }
```


radius nas section
```
ip 172.17.62.206
secret test123
```

radius user section
```
username control_linux
password 123456789
password-type Cleartext-Password
radreply Framed-IP-Address 11.11.11.8
```

radius radacct log
```
Tue Jul 16 12:18:56 2024
        Acct-Status-Type = Stop
        Acct-Session-Id = "1721103102-2"
        NAS-Port-Type = Virtual
        Service-Type = Framed-User
        NAS-Port = 2
        NAS-Port-Id = "eap-public"
        NAS-IP-Address = 172.17.62.206
        Called-Station-Id = "172.17.62.206[4500]"
        Calling-Station-Id = "116.147.11.147[16679]"
        User-Name = "control_linux"
        Framed-IP-Address = 11.11.11.8
        Acct-Output-Octets = 0
        Acct-Output-Packets = 0
        Acct-Input-Octets = 0
        Acct-Input-Packets = 0
        Acct-Session-Time = 424
        Acct-Terminate-Cause = User-Request
        NAS-Identifier = "strongSwan"
        Event-Timestamp = "Jul 16 2024 12:18:56 CST"
        Acct-Unique-Session-Id = "0d7656fc5e59d9398f51e54f5ab21ae7"
        Timestamp = 1721103536

Tue Jul 16 12:18:58 2024
        Acct-Status-Type = Start
        Acct-Session-Id = "1721103102-5"
        NAS-Port-Type = Virtual
        Service-Type = Framed-User
        NAS-Port = 5
        NAS-Port-Id = "eap-public"
        NAS-IP-Address = 172.17.62.206
        Called-Station-Id = "172.17.62.206[4500]"
        Calling-Station-Id = "116.147.11.147[16679]"
        User-Name = "control_linux"
        Framed-IP-Address = 11.11.11.8
        NAS-Identifier = "strongSwan"
        Event-Timestamp = "Jul 16 2024 12:18:58 CST"
        Acct-Unique-Session-Id = "0b5aa867b135516c0eea7ae95f270f87"
        Timestamp = 1721103538
```

**Comment by TelDragon** [comment]
There is another small issue among them.

1. Pools documentation states that multiple pools can be written
pools = radius, local_pools
It still reads from my local address pool. Isn't it in radius priority order


2. Because I have set up multiple logins for the same account
unique = never
How to allocate IP addresses by ID?

```
local {
            auth = eap
            eap_id = control_linux
            id = 14462
        }
```


```
local {
            auth = eap
            eap_id = control_linux
            id = 14463
        }
```

```
local {
            auth = eap
            eap_id = control_linux
            id = 14464
        }
```

```
secrets {
   eap-control_linux {
      id = control_linux
      secret = DDzhMRk5Mp.L
   }
}
```
I don't seem to have obtained ID=14462 in the radius message

**Reply by tobiasbrunner** [reply]
> 1. Pools documentation states that multiple pools can be written
>        pools = radius, local_pools
>        It still reads from my local address pool. Isn't it in radius priority order

No, the order is determined by the plugin order. So you have to [prioritize](https://docs.strongswan.org/docs/5.9/plugins/pluginLoad.html#_runtime_plugin_configuration) the `eap-radius` plugin over the `vici` plugin if you list both pools and want to use RADIUS assigned IPs first.


> 2\. Because I have set up multiple logins for the same account
>         unique = never
>         How to allocate IP addresses by ID?
> I don't seem to have obtained ID=14462 in the radius message

You configured the same EAP-Identity for all of these. The IKE identity will not be used if there is an EAP-Identity. Please also read [this](https://docs.strongswan.org/docs/5.9/plugins/eap-radius.html#_identity_matching).
